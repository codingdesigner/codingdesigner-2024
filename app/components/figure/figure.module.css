.figure {
  display: flex;
  margin-block-end: unset;
  margin-block-start: unset;
  margin-inline-end: unset;
  margin-inline-start: unset;
  padding: var(--spacing-baseline);
  width: fit-content;
}

.bottomBorder {
  composes: figure;
  border-bottom: 1px solid light-dark(var(--color-grayscale-gray6), var(--color-grayscale-black));
}

@container (min-width: 924px) {
  .centerFigure {
    composes: figure;
    margin-left: auto;
    margin-right: auto;
  }
}

.imageFirst {
  composes: figure;
  flex-direction: column;
}
.captionFirst {
  composes: figure;
  flex-direction: column-reverse;
}

.circleCrop {
  border-radius: 50%;
}

.grayScale {
  filter: grayscale(1) contrast(1.25);
}

.figCaption {
  color: light-dark(var(--color-grayscale-gray2), var(--color-grayscale-gray5));
}

.img {
  box-shadow:
    2px 2px 8px light-dark(var(--color-grayscale-gray5), var(--color-grayscale-black)),
    -2px -2px 8px light-dark(var(--color-grayscale-gray5), var(--color-grayscale-black));
  height: unset;
}

.dialog {
  padding: var(--spacing-2);
  border-width: 1px;
  border-style: solid;
  border-color: light-dark(var(--color-grayscale-gray3), var(--color-grayscale-gray2));
  position: relative;
}

.dialogImage {
  border: 1px solid light-dark(var(--color-grayscale-gray5), var(--color-grayscale-gray2));
  height: unset;
}

.dialogClose {
  cursor: pointer;
  position: absolute;
  right: 0;
  top: 2px;
  z-index: 2;
  background-color: transparent;
  border: none;
}

.dialogCloseButton {
  height: 40px;
  width: 40px;
}

.dialogCloseButton circle {
  fill: light-dark(var(--color-grayscale-gray1), var(--color-grayscale-gray3))
}

.dialogCloseButton path {
  fill: light-dark(var(--color-grayscale-white), var(--color-grayscale-black))
}
