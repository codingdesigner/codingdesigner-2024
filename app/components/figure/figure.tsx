'use client';
import React from 'react';
import Image from 'next/image';

import styles from './figure.module.css';

/**
 * Represents the properties for the Figure component.
 */
export interface FigureProps {
  altText?: string; // The alternative text for the image.
  caption?: string; // The caption for the figure.
  imageFile?: {
    height: number;
    src: string; // The image source file path.
    width: number; // The width of the image.
  };
  modalImageFile?: {
    height: number;
    src: string; // The image source file path.
    width: number; // The width of the image.
  } | null;
  imageFirst?: boolean // Render direction of the figure
  centerFigure?: boolean // Whether to center the figure
  circleCrop?: boolean // Whether to center the figure
  grayScale?: boolean // Whether to center the figure
}

export function Figure({
  imageFile,
  altText = "",
  caption = "",
  imageFirst = true,
  centerFigure = false,
  circleCrop = false,
  grayScale = false,
  modalImageFile
}: FigureProps) {
  // Validate the imageFile object
  if (!imageFile || !imageFile.src || !imageFile.width || !imageFile.height) {
    console.error('Invalid imageFile object:', imageFile);
    throw new Error('Invalid imageFile object');
  }

  // Determine the CSS class based on the imageFirst prop
  const figureClass = imageFirst ? styles.imageFirst : styles.captionFirst;
  const justifyClass = centerFigure ? styles.centerFigure : "";
  const circleClass = circleCrop ? styles.circleCrop : "";
  const grayscaleClass = grayScale ? styles.grayScale : "";
  const bottomBorderClass = circleCrop ? "" : styles.bottomBorder;

  // Handler for opening the modal dialog
  const handleDialog = (e) => {
    const dialog = e.target.parentNode.querySelector("dialog");
    if (dialog) dialog.showModal();
  };

  return (
    <figure className={`${figureClass} ${justifyClass} ${bottomBorderClass}`}>
      <Image
        className={`${styles.img} ${circleClass} ${grayscaleClass}`}
        src={imageFile.src}
        width={imageFile.width}
        height={imageFile.height}
        alt={altText || ''}
        onClick={modalImageFile ? handleDialog : undefined}
      />
      {modalImageFile && (
        <dialog className={styles.dialog}>
          <form method="dialog">
            <button className={styles.dialogClose} type="submit">X</button>
            <Image
              className={styles.dialogImage}
              src={modalImageFile.src}
              width={modalImageFile.width}
              height={modalImageFile.height}
              alt={altText || ''}
            />
          </form>
        </dialog>
      )}
      {caption && (
        <figcaption className={styles.figCaption}>{caption || ''}</figcaption>
      )}
    </figure>
  );
}
