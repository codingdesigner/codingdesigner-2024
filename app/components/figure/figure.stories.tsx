import React from 'react';
import type {Meta, StoryObj} from '@storybook/react';
import brandlessCard from "../../static/cards/card-brandless.png"

import {Figure} from './figure';

const meta: Meta<typeof Figure> = {
  title: 'Portfolio/Figure',
  component: Figure,
};

export default meta;

type Story = StoryObj<typeof Figure>;

export const Basic: Story = {
  args: {
    imageFile: brandlessCard,
    altText: "some alt text",
    caption: "A basic caption"
  }
};
