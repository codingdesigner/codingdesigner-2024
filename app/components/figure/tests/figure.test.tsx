import React from 'react';

import {figure} from '../figure';

describe('<figure />', () => {});
