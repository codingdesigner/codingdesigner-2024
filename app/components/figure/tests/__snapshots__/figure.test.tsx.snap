// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Figure /> renders figure unchanged 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <figure
        class="imageFirst  bottomBorder "
      >
        <img
          alt="some alt text"
          class="img  "
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <dialog
          class="dialog"
          data-testid="dialog"
        >
          <form
            method="dialog"
          >
            <button
              class="dialogClose"
              type="submit"
            >
              <svg
                class="dialogCloseButton"
                fill="none"
                height="40"
                viewBox="0 0 40 40"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="20"
                  cy="20"
                  fill="#333333"
                  r="20"
                />
                <path
                  d="M19.778 18.364L26.142 12L27.556 13.414L21.192 19.778L27.556 26.142L26.142 27.556L19.778 21.192L13.414 27.556L12 26.142L18.364 19.778L12 13.414L13.414 12L19.778 18.364Z"
                  fill="white"
                />
              </svg>
            </button>
            <img
              alt="some alt text"
              class="dialogImage"
              data-nimg="1"
              data-testid="modal-image"
              decoding="async"
              height="40"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
              style="color: transparent;"
              width="40"
            />
          </form>
        </dialog>
        <figcaption
          class="figCaption"
          data-testid="figcaption"
        >
          A basic caption
        </figcaption>
      </figure>
    </div>
  </body>,
  "container": <div>
    <figure
      class="imageFirst  bottomBorder "
    >
      <img
        alt="some alt text"
        class="img  "
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <dialog
        class="dialog"
        data-testid="dialog"
      >
        <form
          method="dialog"
        >
          <button
            class="dialogClose"
            type="submit"
          >
            <svg
              class="dialogCloseButton"
              fill="none"
              height="40"
              viewBox="0 0 40 40"
              width="40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="20"
                cy="20"
                fill="#333333"
                r="20"
              />
              <path
                d="M19.778 18.364L26.142 12L27.556 13.414L21.192 19.778L27.556 26.142L26.142 27.556L19.778 21.192L13.414 27.556L12 26.142L18.364 19.778L12 13.414L13.414 12L19.778 18.364Z"
                fill="white"
              />
            </svg>
          </button>
          <img
            alt="some alt text"
            class="dialogImage"
            data-nimg="1"
            data-testid="modal-image"
            decoding="async"
            height="40"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </form>
      </dialog>
      <figcaption
        class="figCaption"
        data-testid="figcaption"
      >
        A basic caption
      </figcaption>
    </figure>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
